File: app/src/main/java/me/pedroguimaraes/weatherdemo/ui/weather/WeatherActivity.java
package me.pedroguimaraes.weatherdemo.ui.weather;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import butterknife.BindView;
import butterknife.ButterKnife;
import me.pedroguimaraes.weatherdemo.R;
import me.pedroguimaraes.weatherdemo.interactors.permissions.PermissionEnforcer;
import me.pedroguimaraes.weatherdemo.model.WeatherInfo;

public class WeatherActivity extends AppCompatActivity implements WeatherContract.WeatherView {

    private WeatherPresenter weatherPresenter;
    private PermissionEnforcer permissionEnforcer;

    @BindView(R.id.srl_weather)
    SwipeRefreshLayout swipeRefreshWeather;

    @BindView(R.id.tv_city_temperature)
    TextView cityTemperature;

    @BindView(R.id.tv_summary)
    TextView citySummary;

    @BindView(R.id.tv_weather_rain)
    TextView rainProbability;

    @BindView(R.id.tv_weather_wind)
    TextView windSpeed;

    @BindView(R.id.tv_city_name)
    TextView cityName;

    @BindView(R.id.img_weather_icon)
    ImageView weatherIcon;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        permissionEnforcer = new PermissionEnforcer(this);
        permissionEnforcer.requestPermissions();

        weatherPresenter = new WeatherPresenter();
        weatherPresenter.attachView(this, new WeatherInteractor());
        ButterKnife.bind(this);


        swipeRefreshWeather.setOnRefreshListener(this::getCurrentWeather);
    }

    @Override
    protected void onResume() {
        super.onResume();
        getCurrentWeather();
    }

    private void getCurrentWeather() {
        if (permissionEnforcer.allPermissionsGranted()) {
            weatherPresenter.onResume();
        }
    }

    @Override
    public void showProgress() {
        swipeRefreshWeather.setRefreshing(true);
    }

    @Override
    public void hideProgress() {
        swipeRefreshWeather.setRefreshing(false);
    }

    @Override
    public void showMessage(String message) {
        hideProgress();
        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
    }

    @Override
    public void setCurrentWeather(WeatherInfo weatherInfo) {
        cityTemperature.setText(weatherInfo.getTemperature());
        citySummary.setText(weatherInfo.getSummary());
        rainProbability.setText(weatherInfo.getRainProbability());
        windSpeed.setText(weatherInfo.getWindSpeed());
        cityName.setText(weatherInfo.getCityName());
        weatherIcon.setImageResource(weatherInfo.getIcon());
    }

    @Override
    protected void onDestroy() {
        weatherPresenter.detachView();
        super.onDestroy();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        permissionEnforcer.enforcePermissions();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == PermissionEnforcer.APP_SETTINGS_REQUEST_CODE) {
            permissionEnforcer.enforcePermissions();
        }
        super.onActivityResult(requestCode, resultCode, data);
    }
}

File: app/src/main/java/me/pedroguimaraes/weatherdemo/ui/weather/WeatherActivity.kt
package me.pedroguimaraes.weatherdemo.ui.weather

import android.Manifest
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import com.tbruyelle.rxpermissions2.RxPermissions
import kotlinx.android.synthetic.main.activity_main.srl_weather
import kotlinx.android.synthetic.main.layout_weather_today.img_weather_icon
import kotlinx.android.synthetic.main.layout_weather_today.tv_city_name
import kotlinx.android.synthetic.main.layout_weather_today.tv_city_temperature
import kotlinx.android.synthetic.main.layout_weather_today.tv_summary
import kotlinx.android.synthetic.main.layout_weather_today.tv_weather_rain
import kotlinx.android.synthetic.main.layout_weather_today.tv_weather_wind
import me.pedroguimaraes.weatherdemo.R
import me.pedroguimaraes.weatherdemo.injection.DependencyInjection
import me.pedroguimaraes.weatherdemo.model.WeatherInfo
import me.pedroguimaraes.weatherdemo.util.goToAppSettings

class WeatherActivity : AppCompatActivity(), WeatherContract.WeatherView {

    private lateinit var weatherPresenter: WeatherPresenter
    private lateinit var rxPermissions: RxPermissions

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val dependencyInjection = DependencyInjection(this)

        rxPermissions = RxPermissions(this)

        weatherPresenter = dependencyInjection.weatherPresenter()
        weatherPresenter.attachView(this)


        srl_weather.setOnRefreshListener { withLocationPermission { weatherPresenter.getWeatherData() } }

        withLocationPermission { weatherPresenter.getWeatherData() }
    }


    override fun showProgress() {
        srl_weather.isRefreshing = true
    }

    override fun hideProgress() {
        srl_weather.isRefreshing = false
    }

    override fun showMessage(message: Int) {
        hideProgress()
        Snackbar.make(srl_weather, message, Snackbar.LENGTH_LONG).show()
    }

    override fun setCurrentWeather(weatherInfo: WeatherInfo) {
        tv_city_temperature.text = weatherInfo.getTemperature()
        tv_summary.text = weatherInfo.summary
        tv_weather_rain.text = weatherInfo.getRainProbability()
        tv_weather_wind.text = weatherInfo.getWindSpeed()
        tv_city_name.text = weatherInfo.getCityName()
        img_weather_icon.setImageResource(weatherInfo.getWeatherIcon()!!)
    }

    override fun onDestroy() {
        weatherPresenter.detachView()
        super.onDestroy()
    }

    private fun withLocationPermission(callback: () -> Unit) {
        this.let { activity ->
            RxPermissions(activity)
                    .requestEach(Manifest.permission.ACCESS_FINE_LOCATION)
                    .subscribe { permission ->
                        when {
                            permission.granted -> { callback() }
                            permission.shouldShowRequestPermissionRationale -> {
                                showPermissionRationale(R.string.permission_location_rationale) { weatherPresenter.getWeatherData() }
                            }
                            else -> {
                                showPermissionRationale(R.string.permission_location_rationale) { goToAppSettings() }
                            }
                        }

                    }
        }
    }

    private fun showPermissionRationale(permissionMessage: Int, method: () -> Unit) {
        val dialogBuilder = AlertDialog.Builder(this)
                .setCancelable(false)
                .setMessage(permissionMessage)
                .setPositiveButton(android.R.string.ok) { _, _ -> withLocationPermission { method() } }
                .setNegativeButton(android.R.string.no) { _, _ -> }

        dialogBuilder.create().show()
    }
}
